File: 325-maximum-size-subarray-sum-equals-k/325-maximum-size-subarray-sum-equals-k.py
class Solution:
    def maxSubArrayLen(self, nums: List[int], k: int) -> int:
        prefixSum = 0
        res = 0
        indices = {}
        
        for i,value in enumerate(nums):
            prefixSum += value
            
            if prefixSum == k:
                res = i + 1
            if prefixSum - k in indices:
                res = max(res,i - indices[prefixSum - k])
            if prefixSum not in indices:
                indices[prefixSum] = i
        return res
        

File: 325-maximum-size-subarray-sum-equals-k/325-maximum-size-subarray-sum-equals-k.cpp
class Solution {
public:
    int maxSubArrayLen(vector<int>& nums, int k) {
        int len = 0;
        long long int cur = 0;
        unordered_map<int,int> map;
        for(int i = 0; i < nums.size(); i++){
            cur += nums[i];
            if(cur == k){
                len = i + 1;
            }
            if(map.find(cur - k) != map.end()){
                len = max(len,i - map[cur - k]);
            }
            if(map.find(cur) == map.end())
                map[cur] = i;
        }
        return len;
    }
};